ee/fleetctl/updates.go

```diff
@@ -320,6 +316,238 @@ func updatesTimestampFunc(c *cli.Context) error {
+		
+role must be one of ['root', 'targets', 'timestamp', 'snapshot']
+		`,
+		Flags:  updatesFlags(),
+		Action: updatesRotateFunc,
+	}
+}
+
+func updatesRotateFunc(c *cli.Context) error {
+	if c.NArg() != 1 {
+		return errors.New("role must be provided")
+	}
+	role := c.Args().Get(0)
+
+	repoPath := c.String("path")
+	repo, err := openRepo(repoPath)
+	if err != nil {
+		return err
+	}
+	store, err := openLocalStore(repoPath)
+	if err != nil {
+		return err
+	}
+
+	if err := checkKeys(repoPath,
+		"root",
+		"targets",
+		"snapshot",
+		"timestamp",
+	); err != nil {
+		return err
+	}
+
+	// Get old keys for role
+	keys, err := store.GetSigningKeys(role)
+	if err != nil {
+		return errors.Wrap(err, "get keys for role")
+	}
+
+	// Prepare to roll back in case of error.
+	success := false
+	commit, rollback, err := prepareCommitRollback(repoPath)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if success {
+			if err := commit(); err != nil {
+				fmt.Println("Warning: failure during commit:", err)
+			}
+		} else {
+			fmt.Println("Rolling back changes.")
+			if err := rollback(); err != nil {
+				fmt.Println("Warning: failure during rollback:", err)
+			}
+		}
+	}()
+
+	// Delete old keys for role
+	for _, key := range keys {
+		id := key.IDs()[0]
+		err := repo.RevokeKeyWithExpires(role, id, time.Now().Add(rootExpirationDuration))
+		if err != nil {
+			// go-tuf keeps keys around even after they are revoked from the manifest. We can skip
+			// tuf.ErrKeyNotFound as these represent keys that are not present in the manifest and
+			// so do not need to be revoked.
+			if !errors.As(err, &tuf.ErrKeyNotFound{}) {
+				return errors.Wrap(err, "revoke key")
+			}
+		}
+	}
+
+	// TODO change passphrase for new key:
+	// Waiting on https://github.com/theupdateframework/go-tuf/pull/163
+
+	// Generate new key for role
+	if err := updatesGenKey(repo, role); err != nil {
+		return err
+	}
+
+	// Re-sign the root metadata
+	if err := repo.Sign("root.json"); err != nil {
+		return errors.Wrap(err, "sign root.json")
+	}
+
+	// Generate new metadata for each role (technically some of these may not need regeneration
+	// depending on which key was rotated, but there should be no harm in generating new ones for each).
+	if err := repo.AddTargetsWithExpires(nil, nil, time.Now().Add(targetsExpirationDuration)); err != nil {
+		return errors.Wrap(err, "generate targets")
+	}
+
+	if err := repo.SnapshotWithExpires(time.Now().Add(snapshotExpirationDuration)); err != nil {
+		return errors.Wrap(err, "generate snapshot")
+	}
+
+	if err := repo.TimestampWithExpires(time.Now().Add(timestampExpirationDuration)); err != nil {
+		return errors.Wrap(err, "generate timestamp")
+	}
+
+	// Commit the changes.
+	if err := repo.Commit(); err != nil {
+		return errors.Wrap(err, "commit repo")
+	}
+
+	success = true
+	return nil
+}
+
+func prepareCommitRollback(repoPath string) (commit, rollback func() error, err error) {
```

Comment:
I think this is a great idea. Not sure how I feel about the name, since it took me a sec to understand what was happening.

Maybe `beginBackupKeysPseudoTx()` ?